state/compatibility_test.go

```diff
@@ -0,0 +1,225 @@
+package state_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	abciv1beta1 "github.com/cometbft/cometbft/api/cometbft/abci/v1beta1"
+	abciv1beta2 "github.com/cometbft/cometbft/api/cometbft/abci/v1beta2"
+	cryptov1 "github.com/cometbft/cometbft/api/cometbft/crypto/v1"
+	statev1 "github.com/cometbft/cometbft/api/cometbft/state/v1"
+	statev1beta2 "github.com/cometbft/cometbft/api/cometbft/state/v1beta2"
+	statev1beta3 "github.com/cometbft/cometbft/api/cometbft/state/v1beta3"
+	typesv1beta1 "github.com/cometbft/cometbft/api/cometbft/types/v1beta1"
+	typesv1beta2 "github.com/cometbft/cometbft/api/cometbft/types/v1beta2"
+)
+
+// Compatibility test across different state proto versions
+
+// This test unmarshal a v1beta2.ABCIResponses as a statev1.LegacyABCIResponses
+// This logic is important for the LoadFinalizeBlockResponse method in the state store
+// The conversion should not fail because they are compatible.
+func TestStateProtoV1Beta2ToV1(t *testing.T) {
+	v1beta2ABCIResponses := newV1Beta2ABCIResponses()
+
+	v1b2Resp, err := v1beta2ABCIResponses.Marshal()
+	require.NoError(t, err)
+	require.NotNil(t, v1b2Resp)
+
+	// un-marshall a v1beta2 ABCI Response as a LegacyABCIResponse
+	legacyABCIResponse := new(statev1.LegacyABCIResponses)
+	err = legacyABCIResponse.Unmarshal(v1b2Resp)
+	require.NoError(t, err)
+	require.NotNil(t, legacyABCIResponse.DeliverTxs)
+	require.NotNil(t, legacyABCIResponse.BeginBlock)
+	require.NotNil(t, legacyABCIResponse.EndBlock)
+	require.Equal(t, 1, len(legacyABCIResponse.DeliverTxs))
+	require.Equal(t, 1, len(legacyABCIResponse.BeginBlock.Events))
+	require.Equal(t, 1, len(legacyABCIResponse.EndBlock.Events))
+	require.Equal(t, int64(100000), legacyABCIResponse.EndBlock.ConsensusParamUpdates.Block.MaxBytes)
+	require.Equal(t, int64(10000), legacyABCIResponse.EndBlock.ConsensusParamUpdates.Evidence.MaxBytes)
+	require.Equal(t, []string{"ed25519"}, legacyABCIResponse.EndBlock.ConsensusParamUpdates.Validator.PubKeyTypes)
+	require.Equal(t, uint64(10), legacyABCIResponse.EndBlock.ConsensusParamUpdates.Version.App)
+}
+
+// This test unmarshal a v1beta2.ABCIResponses as a v1.FinalizeBlockResponse
+// This logic is important for the LoadFinalizeBlockResponse method in the state store
+// The conversion should fail because they are not compatible.
+func TestStateV1Beta2ABCIResponsesAsStateV1FinalizeBlockResponse(t *testing.T) {
+	v1beta2ABCIResponses := newV1Beta2ABCIResponses()
+	data, err := v1beta2ABCIResponses.Marshal()
+	require.NoError(t, err)
+	require.NotNil(t, data)
+
+	// This cannot work since they have different schemas, a wrong wireType error is generated
+	finalizeBlockResponse := new(abci.FinalizeBlockResponse)
+	err = finalizeBlockResponse.Unmarshal(data)
+	require.Error(t, err)
+	require.ErrorContains(t, err, "unexpected EOF")
```

Comment:
I don't understand, by looking at the production changes below, I had understood that the bug is that this unmarshalling actually doesn't error, but leaves some fields unfilled...